### PR TITLE
Alerting: Stop persisting user-defined templates to disk

### LINKED ISSUE #76476: Alerting: Remove disk sync code for templates
Prometheus used to require that templates go on disk to pass them to alertmanager. Grafana has some logic to do this, it syncs all the templates to disk on every AM restart.

Prometheus removed this requirement, we can now pass templates to it from memory instead and use it like a library: https://github.com/prometheus/alertmanager/pull/3225

 We can remove the templates from disk entirely and delete this logic. This gives us faster, more stable config reloads.
Mimir is doing the same thing: https://github.com/grafana/mimir/pull/4967

### PR DESCRIPTION
**Pull request by JacobsonMT**
**What is this feature?**

Updates Grafana Alertmanager to work with new interface from https://github.com/grafana/alerting/pull/161. This change stops passing user-defined templates to the Grafana Alertmanager by persisting them to disk and instead passes them by string.

**Why do we need this feature?**

This PR is a general improvement and simplification, but there are also some specific reasons why we might want to do this:

- Incremental step towards abstracting data layer for Grafana's Alertmanager interface away from needing disk-based storage. Next step would likely focus on `FileStore`.
- Reduces likelihood of consistency complexities on save/apply, especially in the context of future Remote AM work.
- Since `DefaultTemplateString` is now treated like other defaults and therefore consistently applied first. User-defined templates can now rely on being able to always override defaults.

**Who is this feature for?**

Developers mostly, though alert users wanting to consistently override default templates will see improvement here.

**Which issue(s) does this PR fix?**:

Fixes #76476

**Special notes for your reviewer:**

- https://github.com/grafana/alerting/pull/161 needs to be merged first and the change to go.mod updated with the new ref.
- I chose not to remove `Config.Templates` entirely. We should no longer be using it anywhere in local Grafana AM but it's still possible external AMs could use the filenames somehow. We should make a decision one way or the other eventually, but for now keeping it is the safe bet.
- We used to rely on the disk-persisted templates to decide when a template had changed and update the AM config accordingly. This is no longer possible, so instead we replace the hash of `PostableApiAlertingConfig` with the hash of `PostableUserConfig`, which includes the template contents. This change also necessitates we modify `GetStatus` to expect `PostableUserConfig` as the config.

### CODE REVIEW

File: pkg/services/ngalert/notifier/status.go [hunk 1]
```diff
@@ -8,15 +8,15 @@ import (
...
-		return *apimodels.NewGettableStatus(config)
+		return *apimodels.NewGettableStatus(&config.AlertmanagerConfig)
 	}
 
 	if err := json.Unmarshal(status, config); err != nil {
 		am.logger.Error("Unable to unmarshall alertmanager config", "Err", err)
 	}
 
-	return *apimodels.NewGettableStatus(config)
+	return *apimodels.NewGettableStatus(&config.AlertmanagerConfig)
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/notifier/status.go:21]: This does not seem to be related to the PR. Is this the fix that makes it parse the correct format?

**Review by yuri-tceretian — APPROVED** [review]
LGTM.
I am not 100% sure about the ignoring of "Templates" field, though. In Mimir and GMA it's certainly redundant but that's the AM's contract.

File: pkg/services/ngalert/notifier/status.go (hunk 1 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/notifier/status.go:21]: It is related, but not obviously so. I'll try to make it more obvious in a future PR though. Currently, `GrafanaAlertmanager.GetStatus()` just returns the raw config you initially provided to it. It was `PostableApiAlertingConfig` but now it's a `PostableUserConfig`.

**Comment by JacobsonMT** [comment]
> LGTM. I am not 100% sure about the ignoring of "Templates" field, though. In Mimir and GMA it's certainly redundant but that's the AM's contract.

This should be a no-op from my understanding, since we didn't read paths from the `Templates` field before either. I agree that we should consider whether we want to allow that in the future. Unless you meant that this was allowed before and I've broken it somehow?

**Comment by yuri-tceretian** [comment]
AFAIU, the vanilla Alertmanager uses Templates to control what files to read from the templates directory. So the field Template can potentially contain a subset of TemplateFiles. In Mimir and GMA they are always equal because it's hardcoded to be so. 
Making it empty or completely ignoring it, we kinda make it work differently, which is fine by me.

**Review by santihernandezc — APPROVED** [review]
LGTM 👍

File: pkg/services/ngalert/provisioning/templates.go [hunk 2]
```diff
@@ -63,11 +63,6 @@ func (t *TemplateService) SetTemplate(ctx context.Context, orgID int64, tmpl def
 		revision.cfg.TemplateFiles = map[string]string{}
 	}
 	revision.cfg.TemplateFiles[tmpl.Name] = tmpl.Template
-	tmpls := make([]string, 0, len(revision.cfg.TemplateFiles))
```
Comment (alexweav) [inline pkg/services/ngalert/provisioning/templates.go:66]: AFAIU the config still expects the template file "names" to be listed in the config under `Templates`, even if the templates themselves are passed in memory, or it won't be able to load them. Is something else expanding this back out? (or, did this change in alertmanager?)

File: pkg/services/ngalert/notifier/alertmanager.go [hunk 3]
```diff
@@ -321,38 +320,28 @@ func (am *alertmanager) aggregateInhibitMatchers(rules []config.InhibitRule, amu
...
-		return false, err
-	}
-	cfg.AlertmanagerConfig.Templates = paths
-
-	// If neither the configuration nor templates have changed, we've got nothing to do.
-	if !amConfigChanged && !templatesChanged {
-		am.logger.Debug("Neither config nor template have changed, skipping configuration sync.")
+	// If configuration hasn't changed, we've got nothing to do.
+	configHash := md5.Sum(rawConfig)
+	if am.Base.ConfigHash() == configHash {
```
Comment (alexweav) [inline pkg/services/ngalert/notifier/alertmanager.go:331]: Does changing the level at which the hash operates break remote alertmanagers? (@santihernandezc may know more than me here). Will the remote alertmanager still return the hash of the inner, AM config and not this new outer object (PostableUserConfig), it'll not match up and reload needlessly - is this possible?

Do all implementations return the hash at the right level?

Comment (santihernandezc) [reply pkg/services/ngalert/notifier/alertmanager.go:331]: The hash is computed in Grafana and sent to the remote Alertmanager, when we compare configurations we do it using that same hash we sent.
If all config hashes are derived from the same struct, then nothing should break.
Plus, the integration tests are passing 👌

File: pkg/services/ngalert/provisioning/templates.go (hunk 2 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/provisioning/templates.go:66]: Ah, I think I misunderstood you entirely. If you mean upstream prometheus alertmanager needing the `Templates` file names then I don't think it does, or at least this PR appears to be working as intended. When upstream is ran through main then it relies on the `configCoordinator` to subscribe to config changes and would likely require the `Templates` field, but we don't use main and instead piece together an alertmanager part-by-part in `grafana/alerting` [GrafanaAlertmanager.ApplyConfig](https://github.com/grafana/alerting/blob/75cd41d40a70a95444b3bef75fe03f810c585cf7/notify/grafana_alertmanager.go#L346) and manually call `Run()` on the `Dispatcher` and `Inhibitor`.

File: pkg/services/ngalert/notifier/alertmanager.go (hunk 3 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/notifier/alertmanager.go:331]: The way it's currently implemented, this should not affect the remote alertmanagers as this code is internal to the local alertmanager and does not affect what we store in the database.

However, if this somehow changes then we'd likely want all the hashes to be based on `PostableUserConfig` anyways since it contains `TemplateFiles` which is how we'll be passing templates to the remote alertmanager, correct?

I'll let @santihernandezc confirm though.

Comment (santihernandezc) [reply pkg/services/ngalert/notifier/alertmanager.go:331]: > this should not affect the remote alertmanagers as this code is internal to the local alertmanager and does not affect what we store in the database.

Agree, we don't use the internal Alertmanager's hash for the remote/forked AM, we use what we receive via the MOA.

> if this somehow changes then we'd likely want all the hashes to be based on PostableUserConfig anyways since it contains TemplateFiles which is how we'll be passing templates to the remote alertmanager, correct?

Yep, AFAIU that's the way we currently do it. The hash we send to the remote Alertmanager is the one we get from the database in `ApplyConfig`. This hash is derived from `cmd.AlertmanagerConfiguration` ([permalink](https://github.com/grafana/grafana/blob/2184592174b8ec1ff20381349773fef1533d4b9e/pkg/services/ngalert/store/alertmanager.go#L74C4-L74C15)), which is a marshaled `PostableUserConfig` ([permalink](https://github.com/grafana/grafana/blob/2184592174b8ec1ff20381349773fef1533d4b9e/pkg/services/ngalert/notifier/alertmanager.go#L213-L221)).

**Review by alexweav — APPROVED** [review]
LGTM
